### PR TITLE
Add error-chain throughout the codebase

### PR DESCRIPTION
**Pull request by Michael-F-Bryan**
To help with #14 (better error messages), I've included `error-chain` as a dependency and updated almost all the places we return a `Result` to use the new error type instead of `Box<Error>` or `std::io::Error`.

I skipped the handlebars helpers though. Handlebars expects you to return a specific error type, so you can't use the `error-chain` error type there.

Now we just need to let people know to use the new error type when returning a `Result`, and that they should use `chain_err()` to return an error while also attaching a human-readable explanation of what went wrong.

### CODE REVIEW

**Comment by Michael-F-Bryan** [comment]
Travis just errored on the `x86_64-unknown-linux-musl` build with a rather interesting error message. Apparently while trying to compile `backtrace-sys` (a dependency of `error-chain`) the build script failed with:

```
configure: error: in `/home/travis/build/azerupi/mdBook/target/x86_64-unknown-linux-musl/debug/build/backtrace-sys-204dc57c91e9a514/out':
configure: error: C compiler cannot create executables
```

This may be an issue in the environment travis uses because it happened on stable, beta, and nightly for just that platform. For now I'll see if excluding the `backtrace` feature for that target platform works.

**Comment by Michael-F-Bryan** [comment]
Yay, travis is passing again :) 

@azerupi and @budziq, I had to add a couple things to the `x86_64-unknown-linux-musl` build rules in `travis.yml` to get this to work. `error-chain` depends on `backtrace-sys` and on the default travis image you can't compile C libraries using musl. I had to add the following to `travis.yml` for all `x86_64-unknown-linux-musl` targets so `backtrace-sys` will build properly.

```yml
dist: trusty
addons:
  apt:
    packages: &musl_packages
      - musl
      - musl-dev
      - musl-tools
```

**Comment by azerupi** [comment]
Thanks! Looks good :)

File: src/lib.rs [hunk 1]
```diff
@@ -90,3 +93,21 @@ pub mod utils;
...
+        foreign_links {
+            Io(::std::io::Error);
+            HandlebarsRender(::handlebars::RenderError);
+            HandlebarsTemplate(::handlebars::TemplateError);
+            Utf8(::std::string::FromUtf8Error);
+        }
+
+        errors {
+            Subprocess(message: String, output: ::std::process::Output) {
+                description("A subprocess failed")
```
Comment (messense) [inline src/lib.rs]: This caused `mdbook test` only print on failure

```
[*]: Testing file: "/Users/messense/Projects/PyO3/guide/src/./conversions.md"
An error occured:
A subprocess failed
```

**Comment by budziq** [comment]
*@messense this is serious regression. would you be willing to open an
issue?*

**Pull request merged**